File: backend/app/routers/folder.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional, List
from app.db import folder_dao, history_dao
from app.utils.logger import get_logger
from app.utils.response import ResponseWrapper as R

logger = get_logger(__name__)
router = APIRouter()


class FolderCreateRequest(BaseModel):
    id: str
    name: str
    parent_id: Optional[str] = None
    is_expanded: bool = True


class FolderUpdateRequest(BaseModel):
    name: Optional[str] = None
    parent_id: Optional[str] = None
    is_expanded: Optional[bool] = None


class MoveHistoryRequest(BaseModel):
    task_id: str
    folder_id: Optional[str] = None


class FolderResponse(BaseModel):
    id: str
    name: str
    parent_id: Optional[str]
    is_expanded: bool
    created_at: str
    updated_at: str


@router.post("/folders")
async def create_folder(request: FolderCreateRequest):
    """创建新文件夹"""
    try:
        folder = folder_dao.insert_folder(
            folder_id=request.id,
            name=request.name,
            parent_id=request.parent_id,
            is_expanded=request.is_expanded
        )
        
        if not folder:
            return R.error("Failed to create folder", 400)
        
        folder_data = FolderResponse(
            id=folder.id,
            name=folder.name,
            parent_id=folder.parent_id,
            is_expanded=folder.is_expanded,
            created_at=folder.created_at.isoformat(),
            updated_at=folder.updated_at.isoformat()
        )
        return R.success(folder_data.dict(), "Folder created successfully")
    except Exception as e:
        logger.error(f"Error creating folder: {e}")
        return R.error("Internal server error", 500)


@router.get("/folders")
async def get_all_folders():
    """获取所有文件夹"""
    try:
        folders = folder_dao.get_all_folders()
        folder_list = [
            FolderResponse(
                id=folder.id,
                name=folder.name,
                parent_id=folder.parent_id,
                is_expanded=folder.is_expanded,
                created_at=folder.created_at.isoformat(),
                updated_at=folder.updated_at.isoformat()
            ).dict()
            for folder in folders
        ]
        return R.success(folder_list, "Folders retrieved successfully")
    except Exception as e:
        logger.error(f"Error getting folders: {e}")
        return R.error("Internal server error", 500)


@router.put("/folders/{folder_id}")
async def update_folder(folder_id: str, request: FolderUpdateRequest):
    """更新文件夹"""
    try:
        update_data = {}
        if request.name is not None:
            update_data['name'] = request.name
        if request.parent_id is not None:
            update_data['parent_id'] = request.parent_id
        if request.is_expanded is not None:
            update_data['is_expanded'] = request.is_expanded
        
        folder = folder_dao.update_folder(folder_id, **update_data)
        if not folder:
            return R.error("Folder not found", 404)
        
        folder_data = FolderResponse(
            id=folder.id,
            name=folder.name,
            parent_id=folder.parent_id,
            is_expanded=folder.is_expanded,
            created_at=folder.created_at.isoformat(),
            updated_at=folder.updated_at.isoformat()
        )
        return R.success(folder_data.dict(), "Folder updated successfully")
    except Exception as e:
        logger.error(f"Error updating folder: {e}")
        return R.error("Internal server error", 500)


@router.delete("/folders/{folder_id}")
async def delete_folder(folder_id: str):
    """删除文件夹"""
    try:
        success = folder_dao.delete_folder(folder_id)
        if not success:
            return R.error("Folder not found", 404)
        
        return R.success(None, "Folder deleted successfully")
    except Exception as e:
        logger.error(f"Error deleting folder: {e}")
        return R.error("Internal server error", 500)


@router.post("/folders/move-history")
async def move_history_to_folder(request: MoveHistoryRequest):
    """将历史记录移动到文件夹"""
    try:
        success = history_dao.move_history_to_folder(request.task_id, request.folder_id)
        if not success:
            return R.error("History record or folder not found", 404)
        
        return R.success(None, "History moved successfully")
    except Exception as e:
        logger.error(f"Error moving history: {e}")
        return R.error("Internal server error", 500)

File: BillNote_frontend/src/services/folder.ts
import request from '@/utils/request'
import type { Folder } from '@/store/taskStore'

// 文件夹API响应类型
export interface FolderRecord {
  id: string
  name: string
  parent_id?: string
  is_expanded: boolean
  created_at: string
  updated_at: string
}

// API请求类型
export interface CreateFolderRequest {
  id: string
  name: string
  parent_id?: string
  is_expanded?: boolean
}

export interface UpdateFolderRequest {
  name?: string
  parent_id?: string
  is_expanded?: boolean
}

export interface MoveHistoryRequest {
  task_id: string
  folder_id?: string
}

// 将API响应转换为前端Folder类型
export function convertFolderRecordToFolder(record: FolderRecord): Folder {
  return {
    id: record.id,
    name: record.name,
    parentId: record.parent_id,
    isExpanded: record.is_expanded,
    createdAt: record.created_at,
  }
}

// 将前端Folder类型转换为API请求
export function convertFolderToCreateRequest(folder: Folder): CreateFolderRequest {
  return {
    id: folder.id,
    name: folder.name,
    parent_id: folder.parentId,
    is_expanded: folder.isExpanded,
  }
}

export const folderService = {
  // 获取所有文件夹
  async getAllFolders(): Promise<Folder[]> {
    try {
      const response = await request.get('/folders')
      console.log('获取文件夹API响应:', response)
      
      // request.ts 的响应拦截器已经解包了数据，直接返回的是 data 部分
      if (Array.isArray(response)) {
        console.log('解析文件夹数据:', response)
        return response.map(convertFolderRecordToFolder)
      } else {
        console.error('获取文件夹列表失败: 数据格式错误', response)
        return []
      }
    } catch (error) {
      console.error('获取文件夹列表出错:', error)
      return []
    }
  },

  // 创建文件夹
  async createFolder(folder: Folder): Promise<boolean> {
    try {
      const request_data = convertFolderToCreateRequest(folder)
      await request.post('/folders', request_data)
      console.log(`文件夹创建成功: ${folder.name}`)
      return true
    } catch (error) {
      console.error(`创建文件夹失败 ${folder.name}:`, error)
      return false
    }
  },

  // 更新文件夹
  async updateFolder(folderId: string, updates: UpdateFolderRequest): Promise<boolean> {
    try {
      await request.put(`/folders/${folderId}`, updates)
      console.log(`文件夹更新成功: ${folderId}`)
      return true
    } catch (error) {
      console.error(`更新文件夹失败 ${folderId}:`, error)
      return false
    }
  },

  // 删除文件夹
  async deleteFolder(folderId: string): Promise<boolean> {
    try {
      await request.delete(`/folders/${folderId}`)
      console.log(`文件夹删除成功: ${folderId}`)
      return true
    } catch (error) {
      console.error(`删除文件夹失败 ${folderId}:`, error)
      return false
    }
  },

  // 移动历史记录到文件夹
  async moveHistoryToFolder(taskId: string, folderId?: string): Promise<boolean> {
    try {
      const request_data: MoveHistoryRequest = {
        task_id: taskId,
        folder_id: folderId,
      }
      await request.post('/folders/move-history', request_data)
      console.log(`历史记录移动成功: ${taskId} -> ${folderId || 'root'}`)
      return true
    } catch (error) {
      console.error(`移动历史记录失败 ${taskId}:`, error)
      return false
    }
  },

  // 重命名文件夹
  async renameFolder(folderId: string, newName: string): Promise<boolean> {
    return this.updateFolder(folderId, { name: newName })
  },

  // 切换文件夹展开状态
  async toggleFolder(folderId: string, isExpanded: boolean): Promise<boolean> {
    return this.updateFolder(folderId, { is_expanded: isExpanded })
  },

  // 移动文件夹到新的父文件夹
  async moveFolder(folderId: string, newParentId?: string): Promise<boolean> {
    return this.updateFolder(folderId, { parent_id: newParentId })
  },
}